src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PETypeParameterSymbol.cs

```diff
@@ -235,6 +236,147 @@ private ImmutableArray<TypeWithAnnotations> GetDeclaredConstraintTypes()
+            }
+
+            return tokenDecoder;
+        }
+
+        private TypeWithAnnotations GetConstraintTypeOrDefault(PEModuleSymbol moduleSymbol, MetadataReader metadataReader, MetadataDecoder tokenDecoder, GenericParameterConstraintHandle constraintHandle, ref bool hasUnmanagedModreqPattern)
+        {
+            var constraint = metadataReader.GetGenericParameterConstraint(constraintHandle);
+            var typeSymbol = tokenDecoder.DecodeGenericParameterConstraint(constraint.Type, out bool hasUnmanagedModreq);
+
+            if (typeSymbol.SpecialType == SpecialType.System_ValueType)
+            {
+                // recognize "(class [mscorlib]System.ValueType modreq([mscorlib]System.Runtime.InteropServices.UnmanagedType" pattern as "unmanaged"
+                if (hasUnmanagedModreq)
+                {
+                    hasUnmanagedModreqPattern = true;
+                }
+
+                // Drop 'System.ValueType' constraint type if the 'valuetype' constraint was also specified.
+                if (((_flags & GenericParameterAttributes.NotNullableValueTypeConstraint) != 0))
+                {
+                    return default;
+                }
+            }
+
+            var type = TypeWithAnnotations.Create(typeSymbol);
+            type = NullableTypeDecoder.TransformType(type, constraintHandle, moduleSymbol);
+            type = TupleTypeDecoder.DecodeTupleTypesIfApplicable(type, constraintHandle, moduleSymbol);
+            return type;
+        }
+
+        private static bool? IsNotNullableIfReferenceTypeFromConstraintType(TypeWithAnnotations constraintType, ConsList<PETypeParameterSymbol> inProgress, out bool isNonNullableValueType)
+        {
+            if (!(constraintType.Type is PETypeParameterSymbol typeParameter) ||
+                (object)typeParameter.ContainingSymbol != inProgress.Head.ContainingSymbol ||
+                typeParameter.GetConstraintHandleCollection().Count == 0)
+            {
+                return IsNotNullableIfReferenceTypeFromConstraintType(constraintType, out isNonNullableValueType);
+            }
+
+            bool? isNotNullableIfReferenceType = typeParameter.CalculateIsNotNullableIfReferenceType(inProgress, out isNonNullableValueType);
+
+            if (isNonNullableValueType)
+            {
+                Debug.Assert(isNotNullableIfReferenceType == true);
+                return true;
+            }
+
+            if (constraintType.NullableAnnotation.IsAnnotated() || isNotNullableIfReferenceType == false)
+            {
+                return false;
+            }
+            else if (constraintType.NullableAnnotation.IsOblivious() || isNotNullableIfReferenceType == null)
+            {
+                return null;
+            }
+
+            return true;
+        }
+
+        private bool? CalculateIsNotNullableIfReferenceType(ConsList<PETypeParameterSymbol> inProgress, out bool isNonNullableValueType)
```

Comment:
>CalculateIsNotNullableIfReferenceType [](start = 22, length = 37)

nit: consider making this a local function (single use in `IsNotNullableIfReferenceType` property as far as I can tell)

Reply:
> consider making this a local function (single use in IsNotNullableIfReferenceType property as far as I can tell)

I find it clearer to have it as an instance method because it operates on a different instance than its caller.

---
In reply to: [292138728](https://github.com/dotnet/roslyn/pull/36269#discussion_r292138728) [](ancestors = 292138728)